lib/dry/schema/extensions/json_schema/schema_compiler.rb

```diff
@@ -0,0 +1,182 @@
+          result.merge!(type: "object", properties: keys, required: required.to_a)
+          result
+        end
+
+        # @api private
+        def call(ast)
+          visit(ast)
+        end
+
+        # @api private
+        def visit(node, opts = EMPTY_HASH)
+          meth, rest = node
+          public_send(:"visit_#{meth}", rest, opts)
+        end
+
+        # @api private
+        def visit_set(node, opts = EMPTY_HASH)
+          target = (key = opts[:key]) ? self.class.new : self
+
+          node.map { |child| target.visit(child, opts) }
+
+          return unless key
+
+          target_info = opts[:member] ? {items: target.to_h} : target.to_h
+          type = opts[:member] ? "array" : "object"
+
+          keys.update(key => {**keys[key], type: type, **target_info})
+        end
+
+        # @api private
+        def visit_and(node, opts = EMPTY_HASH)
+          left, right = node
+
+          visit(left, opts)
+          visit(right, opts)
+        end
+
+        # @api private
+        def visit_implication(node, opts = EMPTY_HASH)
+          node.each do |el|
+            visit(el, **opts, required: false)
+          end
+        end
+
+        # @api private
+        def visit_each(node, opts = EMPTY_HASH)
+          visit(node, opts.merge(member: true))
+        end
+
+        # @api private
+        def visit_key(node, opts = EMPTY_HASH)
+          name, rest = node
+
+          if opts.fetch(:required, :true)
+            required << name.to_s
+          else
+            opts.delete(:required)
+          end
+
+          visit(rest, opts.merge(key: name))
+        end
+
+        # @api private
+        def visit_not(node, opts = EMPTY_HASH)
+          _name, rest = node
+
+          visit_predicate(rest, opts)
+        end
+
+        # @api private
+        def visit_predicate(node, opts = EMPTY_HASH)
+          name, rest = node
+          key = opts[:key]
+
+          if name.equal?(:key?)
+            prop_name = rest[0][1]
+            keys[prop_name] = {}
+          else
+            type_opts = fetch_type_opts_for_predicate(name, rest)
+            target = keys[key]
+
+            if target[:type]&.include?("array")
+              target[:items] ||= {}
+              merge_opts!(target[:items], type_opts)
+            else
+              merge_opts!(target, type_opts)
+            end
+          end
+        end
+
+        # @api private
+        def fetch_type_opts_for_predicate(name, rest)
+          type_opts = PREDICATE_TO_TYPE.fetch(name) { EMPTY_HASH }.dup
+          type_opts.transform_values! { |v| v.respond_to?(:call) ? v.call(rest[0][1]) : v }
+          type_opts
+        end
+
+        # @api private
+        def merge_opts!(orig_opts, new_opts)
+          new_type = new_opts[:type]
+          orig_type = orig_opts[:type]
+
+          if orig_type && new_type && orig_type != new_type
+            new_opts[:type] = [orig_type, new_type]
+          end
+
+          orig_opts.merge!(new_opts)
```

Comment:
Looks like we can non-bang `#merge` here or just `orig_opts[:type] = `

Reply:
Not sure I follow, we are mutating the opts here so `merge!` would be required I think.

Reply:
`new_opts` or `orig_opts`? `orig_opts[:type] =` would mutate orig opts and if we should mutate _both_ `orig_opts` & `new_opts` — it's strange. AFAICS it's used as a return value, so mutation could be skipped.

Again, didn't read the overall logic (yet?) and it's just technical comments.